### ISSUE TITLE
Switching to Localhost throws Error

### ISSUE THREAD

**Issue by crazyrabbitLTC** [NoahZinsmeister/web3-react#73]
**Bug Description**
When switching MetaMask to local host, while using an injected connector created thus: 

`export const injectedConnector = new InjectedConnector({ supportedChainIds: [ 1, 3, 4, 5, 10, 42, 77, 99, 100 ] })`

Web3React throws the error. This happens in our own code as well as on the code-sandbox. 
![Screen Shot 2020-02-01 at 11 24 55 AM](https://user-images.githubusercontent.com/1938013/73595343-89e66a00-44e5-11ea-9b64-7f7aebdeefa1.png)

```
Unhandled Rejection (Error): Invariant failed: chainId loading is not an integer
invariant
node_modules/tiny-invariant/dist/tiny-invariant.esm.js:10
normalizeChainId
src/normalizers.ts:13
  10 |   chainId = chainId.replace(/^Ox/, '0x')
  11 | 
  12 |   const parsedChainId = Number.parseInt(chainId, chainId.trim().substring(0, 2) === '0x' ? 16 : 10)
> 13 |   invariant(!Number.isNaN(parsedChainId), `chainId ${chainId} is not an integer`)
     | ^  14 |   return parsedChainId
  15 | } else {
  16 |   invariant(Number.isInteger(chainId), `chainId ${chainId} is not an integer`)
View compiled
(anonymous function)
src/manager.ts:181
  178 | 
  179 | // updates are handled differently depending on whether the connector is active vs in an error state
  180 | if (!error) {
> 181 |   const chainId = update.chainId === undefined ? undefined : normalizeChainId(update.chainId)
      | ^  182 |   if (chainId !== undefined && !!connector.supportedChainIds && !connector.supportedChainIds.includes(chainId)) {
  183 |     const error = new UnsupportedChainIdError(chainId, connector.supportedChainIds)
  184 |     onError ? onError(error) : dispatch({ type: ActionType.ERROR, payload: { error } })
View compiled
▶ 2 stack frames were collapsed.
InjectedConnector.emitUpdate
src/index.ts:27
  24 |   sortProperties,
  25 |   getAttributeMode,
  26 |   getElementDataset,
> 27 |   getIdElementProperties,
  28 |   getDataElementProperties,
  29 |   createAttributeSelector,
  30 | } from '~/lib/utils';
View compiled
InjectedConnector.handleNetworkChanged
src/index.ts:66
  63 | 
  64 | const feature = element.getAttribute(DATA_FEATURE);
  65 | 
> 66 | if (!feature) {
     | ^  67 |   return console.error(`Feature attribute was not added to the element`);
  68 | }
  69 | 
View compiled
▶ 27 stack frames were collapsed.
```

**Reproduction**
[CodeSandbox reproduction]()

Steps to reproduce:
1. 
2. 

**Expected Behavior**
Expected to handle the case of using a local network. 

**Additional Context**

**Comment by crazyrabbitLTC** [comment]
Additionally, I've noticed this error also when switching between the testnet, it seems to depend on how quickly perhaps metamask returns information. That said, perhaps it's not a great feature of metamask to return a string for a value that is otherwise a number.

**Comment by maxaleks** [comment]
Also the same error occurs when adding a custom network in Metamask without specifying the network id. So I think the error is due to the lack of network id

**Comment by treeder** [comment]
> Also the same error occurs when adding a custom network in Metamask without specifying the network id. So I think the error is due to the lack of network id

I see it when I do have the chain id set. 🤷‍♂️

**Comment by leovigna** [comment]
I also have this issue.

**Comment by leovigna** [comment]
Ok so I think I figured out the bug and a quick patch for it. I think when Metamask is connected to a custom network that supports the eth_chainId RPC request, it first sets the chainId to the value set in the form by the user (0x0..), then it makes an RPC request and sets it to 'loading', finally it updates again with the returned networkId (web3-react doesn't distinguish the two and emits and update on each update.

TLDR: Metamask emits chainId: 'loading' which breaks the nomralizer.ts.

Quick patch that you can apply within your code without the need to edit web3-react:
```
export const injected = new InjectedConnector({
  supportedChainIds: SUPPORTED_CHAIN_IDS
})

//Patch networkChanged error
//@ts-ignore
injected.handleNetworkChanged = (networkId: string | number) => {
  console.debug("Handling 'networkChanged' event with payload", networkId)
  if (networkId === "loading") return; //Ignore loading, networkId as causes errors
  //@ts-ignore
  injected.emitUpdate({ chainId: networkId, provider: window.ethereum })
}
```

I can also do a pull request if @NoahZinsmeister is open to it though I don't know if simply ignoring chainId === "loading" is the best longterm fix for this bug.

**Comment by leovigna** [comment]
Also if you want to support Metamask when the user doesn't explicitly set the chainId of the custom network, you can use the patch below. web3-react will still update when Metamask fetches the chainId with an rpc request. Note that this could cause errors if your rpc doesn't support eth_chainId (or eth_networkId, not sure which one Metamask calls).
```ts
//Patch chainChanged 0xNaN
//@ts-ignore
injected.handleChainChanged = (chainId: string | number) => {
  console.debug("Handling 'chainChanged' event with payload", chainId)
  if (chainId === "0xNaN") return; //Ignore 0xNaN, when user doesn't set chainId
  //@ts-ignore
  injected.emitUpdate({ chainId, provider: window.ethereum })
}
```

**Comment by DannyDelott** [comment]
@leovigna this solved my problem, thank you. I did make a slight change so that the original handler is preserved:


```ts
//@ts-ignore
const originalChainIdChangeHandler = injectedConnector.handleChainChanged;
//@ts-ignore
injectedConnector.handleChainChanged = (chainId: string | number) => {
  console.debug("Handling 'chainChanged' event with payload", chainId);
  if (chainId === "0xNaN") {
    return; //Ignore 0xNaN, when user doesn't set chainId
  }
  originalChainIdChangeHandler(chainId);
};
```

**Comment by leovigna** [comment]
Oh yes this is cleaner 👍

**Comment by NoahZinsmeister** [comment]
hi! library author here, sorry for the (extremely) slow responses; for an update on the state of this library, please see https://github.com/NoahZinsmeister/web3-react/issues/163#issuecomment-1010210571

TLDR: v6 is deprecated and going into maintenance mode. i expect not to make any non-security related changes going forward (however, you're always welcome to! the beauty of open source)

i'll be closing a lot of issues with this message right now, but please feel free to re-open if you think the topic is still relevant!

thanks for using web3-react <3
